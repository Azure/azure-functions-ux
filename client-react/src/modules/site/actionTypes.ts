export const AREA_STRING = '@site';
export const UPDATE_RESOURCE_ID = `@site/UPDATE_RESOURCE_ID`;
export const SITE_FETCH_REQUEST = `@site/FETCH_REQUEST`;
export const SITE_FETCH_SUCCESS = `@site/FETCH_SUCCESS`;
export const SITE_FETCH_FAILURE = `@site/FETCH_FAILURE`;
export const SITE_UPDATE_REQUEST = `@site/UPDATE_REQUEST`;
export const SITE_UPDATE_SUCCESS = `@site/UPDATE_SUCCESS`;
export const SITE_UPDATE_FAILURE = `@site/UPDATE_FAILURE`;
