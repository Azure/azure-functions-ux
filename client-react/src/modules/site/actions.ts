import { createStandardAction } from 'typesafe-actions';

import { ArmObj, Site } from '../../models/WebAppModels';
import {
  SITE_FETCH_FAILURE,
  SITE_FETCH_REQUEST,
  SITE_FETCH_SUCCESS,
  SITE_UPDATE_FAILURE,
  SITE_UPDATE_REQUEST,
  SITE_UPDATE_SUCCESS,
  UPDATE_RESOURCE_ID,
} from './actionTypes';

export const updateResourceId = createStandardAction(UPDATE_RESOURCE_ID).map((id: string) => ({ id }));
export const fetchSiteRequest = createStandardAction(SITE_FETCH_REQUEST)();
export const fetchSiteSuccess = createStandardAction(SITE_FETCH_SUCCESS).map((site: ArmObj<Site>) => ({ site }));
export const fetchSiteFailure = createStandardAction(SITE_FETCH_FAILURE).map((error: Error) => ({
  error,
}));

export const updateSiteRequest = createStandardAction(SITE_UPDATE_REQUEST).map((site: ArmObj<Site>) => ({ site }));
export const updateSiteSuccess = createStandardAction(SITE_UPDATE_SUCCESS).map((site: ArmObj<Site>) => ({ site }));
export const updateSiteFailure = createStandardAction(SITE_UPDATE_FAILURE).map((error: Error) => ({
  error,
}));
