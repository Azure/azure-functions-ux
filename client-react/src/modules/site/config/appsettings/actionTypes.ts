export const AREA_STRING = '@appsettings';
export const APP_SETTINGS_FETCH_REQUEST = '@appsettings/FETCH_REQUEST';
export const APP_SETTINGS_FETCH_SUCCESS = '@appsettings/FETCH_SUCCESS';
export const APP_SETTINGS_FETCH_FAILURE = '@appsettings/FETCH_FAILURE';

export const APP_SETTINGS_UPDATE_REQUEST = '@appsettings/UPDATE_REQUEST';
export const APP_SETTINGS_UPDATE_SUCCESS = '@appsettings/UPDATE_SUCCESS';
export const APP_SETTINGS_UPDATE_FAILURE = '@appsettings/UPDATE_FAILURE';
