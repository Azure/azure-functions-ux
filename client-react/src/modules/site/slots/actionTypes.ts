export const AREA_STRING = '@slots';
export const SLOTS_FETCH_REQUEST = '@slots/FETCH_REQUEST';
export const SLOTS_FETCH_SUCCESS = '@slots/FETCH_SUCCESS';
export const SLOTS_FETCH_FAILURE = '@slots/FETCH_FAILURE';

export const SLOTS_UPDATE_REQUEST = '@slots/UPDATE_REQUEST';
export const SLOTS_UPDATE_SUCCESS = '@slots/UPDATE_SUCCESS';
export const SLOTS_UPDATE_FAILURE = '@slots/UPDATE_FAILURE';
