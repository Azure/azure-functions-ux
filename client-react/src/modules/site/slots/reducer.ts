import { combineReducers } from 'redux';
import { ActionType } from 'typesafe-actions';

import { ArmArray, Site } from '../../../models/WebAppModels';
import { metadataReducer } from '../../ApiReducerHelper';
import { ApiState } from '../../types';
import * as actions from './actions';
import { AREA_STRING, SLOTS_FETCH_SUCCESS } from './actionTypes';

export type SlotsAction = ActionType<typeof actions>;
export type SlotsState = ApiState<ArmArray<Site>>;
export const InitialState = {
  data: { value: [] },
};

export default combineReducers<SlotsState, SlotsAction>({
  metadata: metadataReducer(AREA_STRING),
  data: (state = InitialState.data, action) => {
    switch (action.type) {
      case SLOTS_FETCH_SUCCESS:
        return action.slotList;
      default:
        return state;
    }
  },
});
