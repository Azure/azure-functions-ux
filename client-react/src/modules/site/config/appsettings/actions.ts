import { IAction } from '../../../../models/action';
import { IAppSettingsState } from './reducer';

export const UPDATE_SITE_APP_SETTINGS = 'UPDATE_SITE_APP_SETTINGS';
export const updateCurrentSiteAppSettings = (settings: Partial<IAppSettingsState>): IAction<Partial<IAppSettingsState>> => ({
  payload: settings,
  type: UPDATE_SITE_APP_SETTINGS,
});

export const UPDATE_SITE_APP_SETTINGS_NO_CACHE = 'UPDATE_SITE_APP_SETTINGS_NO_CACHE';
export const updateCurrentSiteAppSettingsNoCache = (settings: Partial<IAppSettingsState>): IAction<Partial<IAppSettingsState>> => ({
  payload: settings,
  type: UPDATE_SITE_APP_SETTINGS,
});
