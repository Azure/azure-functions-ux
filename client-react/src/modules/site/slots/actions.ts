import { IAction } from '../../../models/action';
import { ISlotListState } from './reducer';

export const UPDATE_SLOT_LIST = 'UPDATE_SLOT_LIST';
export const updateSlotList = (state: Partial<ISlotListState>): IAction<Partial<ISlotListState>> => ({
  payload: state,
  type: UPDATE_SLOT_LIST,
});

export const UPDATE_SLOT_LIST_NO_CACHE = 'UPDATE_SLOT_LIST_NO_CACHE';
export const updateSlotListNoCache = (state: Partial<ISlotListState>): IAction<Partial<ISlotListState>> => ({
  payload: state,
  type: UPDATE_SLOT_LIST_NO_CACHE,
});
