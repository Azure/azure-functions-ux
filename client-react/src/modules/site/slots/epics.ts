import { combineEpics, Epic } from 'redux-observable';
import { from, of } from 'rxjs';
import { catchError, filter, map, switchMap } from 'rxjs/operators';
import { isOfType } from 'typesafe-actions';

import * as Types from '../../types';
import { fetchSlotsFailure, fetchSlotsSuccess } from './actions';
import { SLOTS_FETCH_REQUEST } from './actionTypes';
import { SlotsAction } from './reducer';

export const fetchSlotsFlow: Epic<SlotsAction, SlotsAction, Types.RootState, Types.Services> = (action$, store, { slotsApi }) =>
  action$.pipe(
    filter(isOfType(SLOTS_FETCH_REQUEST)),
    switchMap(() =>
      from(slotsApi.fetchSlots(store.value)).pipe(
        map(fetchSlotsSuccess),
        catchError(err => of(fetchSlotsFailure(err)))
      )
    )
  );
export default combineEpics(fetchSlotsFlow);
