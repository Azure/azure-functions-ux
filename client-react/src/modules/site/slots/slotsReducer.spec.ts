import mockAxios from 'jest-mock-axios';
import { ActionsObservable } from 'redux-observable';
import { toArray } from 'rxjs/operators';

import rootReducer from '../../';
import { IStartupInfo } from '../../../models/portal-models';
import { ArmArray, ArmObj, Site } from '../../../models/WebAppModels';
import { getStartupInfoAction } from '../../portal/actions';
import { RootState, Services } from '../../types';
import { updateResourceId } from '../actions';
import { fetchSlotsFailure, fetchSlotsRequest, fetchSlotsSuccess } from './actions';
import { SLOTS_FETCH_FAILURE, SLOTS_FETCH_SUCCESS } from './actionTypes';
import { fetchSlotsFlow } from './epics';
import reducer from './reducer';
import slotApiService from './slotsApiService';

describe('Slots Store Epics', () => {
  const successDeps = {
    slotsApi: {
      fetchSlots: async (state: RootState): Promise<ArmArray<Site>> => {
        return { value: [{ name: 'testSlot' } as ArmObj<Site>] };
      },
    },
  } as Services;

  const failDeps = {
    slotsApi: {
      fetchSlots: async (state: RootState): Promise<ArmArray<Site>> => {
        throw new Error('failuremessage');
      },
    },
  } as Services;
  it('Sends Success Action with slots object on Successful Fetch', function(done) {
    let action$ = ActionsObservable.of(fetchSlotsRequest());
    const output$ = fetchSlotsFlow(action$, {} as any, successDeps);
    output$.pipe(toArray()).subscribe(actions => {
      expect(actions.length).toBe(1);
      const action = actions[0];
      expect(action.type).toBe(SLOTS_FETCH_SUCCESS);
      if (action.type === SLOTS_FETCH_SUCCESS) {
        expect(action.slotList.value[0].name).toBe('testSlot');
      }

      done();
    });
  });

  it('Sends Error Action with error on failed Fetch', function(done) {
    let action$ = ActionsObservable.of(fetchSlotsRequest());
    const output$ = fetchSlotsFlow(action$, {} as any, failDeps);
    output$.pipe(toArray()).subscribe(actions => {
      expect(actions.length).toBe(1);
      const action = actions[0];
      expect(action.type).toBe(SLOTS_FETCH_FAILURE);
      if (action.type === SLOTS_FETCH_FAILURE) {
        expect(action.error.message).toBe('failuremessage');
      }
      done();
    });
  });
});

describe('Slots Store Reducer', () => {
  const initialState = reducer(undefined, {} as any);
  describe('initial state', () => {
    it('should match a snapshot', () => {
      expect(initialState).toMatchSnapshot();
    });

    it('loading and updating should be false', () => {
      expect(initialState.metadata.loading).toBe(false);
      expect(initialState.metadata.updating).toBe(false);
    });
  });

  describe('Site Fetch Stories', () => {
    it('should trigger loading the fetch is requested', () => {
      const action = fetchSlotsRequest();
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(true);
    });

    it('metadata and site should update on successful load', () => {
      const action = fetchSlotsSuccess({ value: [{ name: 'testsite' }] } as any);
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(false);
      expect(state.data.value[0].name).toBe('testsite');
    });

    it('error should be reflected on failed load', () => {
      const action = fetchSlotsFailure(new Error('testerror'));
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(false);
      expect(state.metadata.fetchError).toBe(true);
      expect(state.metadata.fetchErrorObject.message).toBe('testerror');
    });
  });
});

describe('Site Service', () => {
  const initialState = rootReducer(undefined, {} as any);
  const catchFn = jest.fn();
  const thenFn = jest.fn();
  let state;
  beforeEach(() => {
    const updateResourceIdAction = updateResourceId('resourceid');
    const updateSUIAction = getStartupInfoAction({
      token: 'testtoken',
      armEndpoint: 'testEndpoint',
    } as IStartupInfo);

    state = rootReducer(rootReducer(initialState, updateResourceIdAction), updateSUIAction);
  });

  afterEach(() => {
    mockAxios.reset();
    thenFn.mockClear();
    catchFn.mockClear();
  });

  it('Fetch Api calls api with appropriate info from production app', async () => {
    const fetcher = slotApiService.fetchSlots(state);
    let siteApiRequestInfo = mockAxios.lastReqGet();
    expect(mockAxios).toHaveBeenCalledWith({
      method: 'GET',
      url: 'testEndpointresourceid/slots?api-version=2018-02-01',
      data: null,
      headers: {
        Authorization: `Bearer testtoken`,
      },
    });
    let responseObj = { data: { value: [{ name: test }] } };
    mockAxios.mockResponse(responseObj, siteApiRequestInfo);
    const result = await fetcher;
    expect(result.value.length).toBe(1);
  });

  it('Fetch Api calls api with appropriate info from slots app', async () => {
    const updateResourceIdAction = updateResourceId('resourceid/slots/slot');
    state = rootReducer(state, updateResourceIdAction);
    const fetcher = slotApiService.fetchSlots(state);
    let siteApiRequestInfo = mockAxios.lastReqGet();
    expect(mockAxios).toHaveBeenCalledWith({
      method: 'GET',
      url: 'testEndpointresourceid/slots?api-version=2018-02-01',
      data: null,
      headers: {
        Authorization: `Bearer testtoken`,
      },
    });
    let responseObj = { data: { value: [{ name: test }] } };
    mockAxios.mockResponse(responseObj, siteApiRequestInfo);
    const result = await fetcher;
    expect(result.value.length).toBe(1);
  });

  it('Fetch Api should throw on error', async () => {
    const fetcher = slotApiService
      .fetchSlots(state)
      .then(thenFn)
      .catch(catchFn);
    mockAxios.mockError(new Error('errorMessage'));
    await fetcher;
    expect(thenFn).not.toHaveBeenCalled();
    expect(catchFn).toHaveBeenCalled();
  });
});
