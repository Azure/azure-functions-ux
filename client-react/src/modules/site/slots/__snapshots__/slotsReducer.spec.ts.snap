// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slots Store Reducer initial state should match a snapshot 1`] = `
Object {
  "data": Object {
    "value": Array [],
  },
  "metadata": Object {
    "fetchError": false,
    "fetchErrorObject": [Error],
    "loading": false,
    "updateError": false,
    "updateErrorObject": [Error],
    "updating": false,
  },
}
`;
